app/src/main/java/me/devcexx/call2alarm/MainActivity.kt

```diff
@@ -83,22 +102,24 @@ class MainActivity : ComponentActivity() {
             verticalArrangement = Arrangement.Top
         ) {
 
-            Preferences()
+            Preferences(permissionsRequest)
         }
     }
 
     @Composable
-    fun Preferences() {
+    fun Preferences(permissionsRequest: MultiplePermissionsState) {
         val (interceptionIsEnabled, _) = interceptionEnabled
         val (ringtoneVolumeSettingValue, _) = ringtoneVolumeSetting
         PreferenceSet(context = applicationContext) {
             PreferenceGroup(titleId = R.string.pref_group_global) {
+                val context = LocalContext.current
                 SwitchPreference(
                     titleId = R.string.pref_enable_call_handling_title,
                     subtitleId = R.string.pref_enable_call_handling_subtitle,
                     state = interceptionEnabled,
                     onChange = {
-                        ensureServiceInCorrectState()
+                        val hasPermissions = permissionsRequest.checkMandatoryPermissions()
+                        context.onPermissionsProvided(hasPermissions && it)
```

Comment:
[nit] This is logically correct, but semantically, you're passing to the parameter hasPermission a value that indicates that the app has the permissions granted _and_ that the switch is on. Not sure if this is clear enough.

Reply:
The previous behaviour is that if no permissions have been granted, then the switch shouldn't be allowed to be switched on. Can we add here some line to keep that logic in that way? (Eventually we might want to add some informative message or whatever, but for now it is fine to just keep it working as it was before)